// @medusajs/js-sdk/dist/esm/admin/views.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
export class Views {
    constructor(client) {
        this.client = client;
    }
    // Generic method to get columns for any entity
    columns(entity, query, headers) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.client.fetch(`/admin/views/${entity}/columns`, {
                method: "GET",
                headers,
                query,
            });
        });
    }
    listConfigurations(entity, query, headers) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.client.fetch(`/admin/views/${entity}/configurations`, {
                method: "GET",
                headers,
                query,
            });
        });
    }
    createConfiguration(entity, body, headers) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.client.fetch(`/admin/views/${entity}/configurations`, {
                method: "POST",
                headers,
                body,
            });
        });
    }
    retrieveConfiguration(entity, id, query, headers) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.client.fetch(`/admin/views/${entity}/configurations/${id}`, {
                method: "GET",
                headers,
                query,
            });
        });
    }
    updateConfiguration(entity, id, body, headers) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.client.fetch(`/admin/views/${entity}/configurations/${id}`, {
                method: "POST",
                headers,
                body,
            });
        });
    }
    deleteConfiguration(entity, id, headers) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.client.fetch(`/admin/views/${entity}/configurations/${id}`, {
                method: "DELETE",
                headers,
            });
        });
    }
    retrieveActiveConfiguration(entity, headers) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.client.fetch(`/admin/views/${entity}/configurations/active`, {
                method: "GET",
                headers,
            });
        });
    }
    setActiveConfiguration(entity, body, headers) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.client.fetch(`/admin/views/${entity}/configurations/active`, {
                method: "POST",
                headers,
                body,
            });
        });
    }
}
