// @perseidesjs/notification-nodemailer/.medusa/server/src/providers/nodemailer/service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodemailerNotificationProviderService = void 0;
const utils_1 = require("@medusajs/framework/utils");
const nodemailer_1 = __importDefault(require("nodemailer"));
class NodemailerNotificationProviderService extends utils_1.AbstractNotificationProviderService {
    constructor(_, options) {
        super();
        this.options_ = options;
        this.transporter = nodemailer_1.default.createTransport(options);
    }
    static validateOptions(options) {
        if (!options.host) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `"host" is required in the ${this.identifier} provider's options.`);
        }
        if (!options.port) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `"port" is required in the ${this.identifier} provider's options.`);
        }
    }
    async send(notification) {
        const { to, content } = notification;
        if (!(0, utils_1.isDefined)(this.options_.from) && !(0, utils_1.isDefined)(notification.from)) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `"from" is required in the ${NodemailerNotificationProviderService.identifier} provider's options.`);
        }
        let html;
        if ((0, utils_1.isDefined)(content?.html) || (0, utils_1.isDefined)(content?.text)) {
            html = content?.html ?? content?.text;
        }
        let attachments = [];
        if ((0, utils_1.isDefined)(notification.attachments) &&
            Array.isArray(notification.attachments) &&
            notification.attachments.length > 0) {
            attachments = notification.attachments.map((attachment) => ({
                content: attachment.content,
                filename: attachment.filename,
                contentType: attachment.content_type,
                contentDisposition: attachment.disposition === 'inline' ||
                    attachment.disposition === 'attachment'
                    ? attachment.disposition
                    : undefined,
                cid: attachment.id,
                contentTransferEncoding: attachment.content_type === 'text/html'
                    ? 'quoted-printable'
                    : 'base64',
            }));
        }
        const mailOptions = {
            from: notification.from ?? this.options_.from,
            to,
            subject: content?.subject,
        };
        if ((0, utils_1.isDefined)(html)) {
            mailOptions.html = html;
        }
        if (attachments.length > 0) {
            mailOptions.attachments = attachments;
        }
        const info = await this.transporter.sendMail(mailOptions);
        return {
            id: info.messageId,
        };
    }
}
exports.NodemailerNotificationProviderService = NodemailerNotificationProviderService;
NodemailerNotificationProviderService.identifier = 'nodemailer';
exports.default = NodemailerNotificationProviderService;
